SALTO-1715: addDeploymentAnnotations now supports inner types

alonstern opened the pull request:
Added support for inner types in `addDeploymentAnnotations`

---
_Release Notes_: 
None

---
_User Notifications_: 
None

netama commented on packages/adapter-components/src/elements/swagger/deployment/annotations.ts:
```diff
@@ -13,49 +13,84 @@
+
+  annotatedTypes.add(type.elemID.getFullName())
+
+  const fields = getFields(
+    swagger,
+    schema
+  )
```
nit:
```suggestion
  const fields = getFields(swagger, schema)
```

netama commented on packages/adapter-components/src/elements/swagger/deployment/annotations.ts:
```diff
@@ -13,49 +13,84 @@
+    if (field === undefined) {
+      return
+    }
+    field.annotations[OPERATION_TO_ANNOTATION[action]] = true
```
isn't true the default?

alonstern replied:
The logic of this function is first, mark everything as false, then mark with true what is deployable. so in this point field.annotations[OPERATION_TO_ANNOTATION[action]] won't be `undefined`

netama commented on packages/adapter-components/src/elements/swagger/deployment/annotations.ts:
```diff
@@ -13,49 +13,84 @@
+      await setTypeAnnotations(fieldType, properties, swagger, action, annotatedTypes)
+    }
+
+    if (isListType(fieldType)) {
```
what about map types? and specifically `additionalProperties`?

netama commented on packages/adapter-components/src/elements/swagger/deployment/annotations.ts:
```diff
@@ -76,19 +111,17 @@ export const addDeploymentAnnotationsFromSwagger = (
     }
 
-    delete type.annotations[OPERATION_TO_ANNOTATION[operation as ActionName]]
+    type.annotations[OPERATION_TO_ANNOTATION[operation as ActionName]] = true
```
why the change?

alonstern replied:
Since it is hidden now, I thought it better to be explicit about it, but it doesn't really matter

alonstern replied:
Changed back

netama commented on packages/adapter-components/src/elements/swagger/deployment/annotations.ts:
```diff
@@ -76,19 +111,17 @@ export const addDeploymentAnnotationsFromSwagger = (
-      }
-    })
+    if (schema === undefined) {
+      return
```
should we log a warning in this case?

netama commented on packages/adapter-components/src/elements/swagger/deployment/annotations.ts:
```diff
@@ -13,49 +13,84 @@
+
+  annotatedTypes.add(type.elemID.getFullName())
+
+  const fields = getFields(
```
nit: maybe `swaggerFields` to make it easier to distinguish from the existing objecttype's fields?

netama commented on packages/adapter-components/src/elements/swagger/deployment/annotations.ts:
```diff
@@ -13,49 +13,84 @@
+    schema
+  )
+
+  await awu(Object.entries(fields)).forEach(async ([name, properties]) => {
```
technically we allow ourselves to add/remove fields with `fieldTypeOverrides` - do we need any handling for these?

alonstern replied:
I don't think so. I guess if we add a new field we will need to also add its annotations

netama replied:
will it be deployable by default in this flow?

alonstern replied:
If the user did not add the annotations for it and it really does not appear in the swagger, then it won't be deployable (because it really isn't)

netama commented on packages/adapter-components/src/elements/swagger/deployment/annotations.ts:
```diff
@@ -100,19 +133,31 @@ export const addDeploymentAnnotationsFromSwagger = (
-  type.annotations[DEPLOYMENT_ANNOTATIONS.UPDATABLE] = false
-  type.annotations[DEPLOYMENT_ANNOTATIONS.DELETABLE] = false
+): Promise<void> => {
+  type.annotations[DEPLOYMENT_ANNOTATIONS.CREATABLE] = Boolean(
```
why the `Boolean()` constructor?

alonstern replied:
if the annotation was already marked `true` manually, I don't want to override it, so I change it for false to use the `Boolean`

netama replied:
but didn't we say our default is also true? I think I'm missing something around when we expect this to be defined now

alonstern replied:
It's not that this function assumes that the annotation is false if it is undefined, it will still go to the swagger and if it finds that it is deployable it will remove the annotation with the false value. It's just that if the annotation were explicitly marked as true before this function ran, we don't want to override it with false (for cases that the field does not appear in the swagger, but we do have custom logic for deploying it).

The default value when the annotation is undefined is set to true by the change validator and by the `deployChange` func.
It is meant mostly for adapters that don't use this function and don't set the annotations at all or only partially, so their current behavior won't change

netama commented on packages/adapter-components/src/elements/swagger/deployment/annotations.ts:
```diff
@@ -100,19 +133,31 @@ export const addDeploymentAnnotationsFromSwagger = (
   })
-  swaggers.forEach(swagger => addDeploymentAnnotationsFromSwagger(type, swagger, endpointDetails))
+  await Promise.all(
+    swaggers.map(swagger => addDeploymentAnnotationsFromSwagger(type, swagger, endpointDetails))
```
I guess this was kind of already here (sorry I missed it in the previous PR), but - 
do we ever expect multiple swaggers to have definitions for the same type?
if yes - what's the expected result? any chance of overlaps for the same field (in which case there is a race condition here that we should probably avoid)?
if no - can we enforce that?

alonstern replied:
Our only example of multiple swaggers currently is in Jira, in which we don't expect the same type to appear more than once. But if so, we will sort of get the wanted outcome, i.e., all the fields that are deployable with at least one of the swaggers will be marked with true, and the rest with false

netama commented on packages/adapter-components/src/elements/swagger/deployment/annotations.ts:
```diff
@@ -122,16 +167,16 @@ const addDeploymentAnnotationsToType = (
-): void => {
-  types.forEach(
+): Promise<void> => {
+  await Promise.all(types.map(
```
another potential race condition here (I think) - multiple types can have the same subtype, what do we expect to happen in that case? and do we really need to parallelize (and make this non-deterministic)?

alonstern replied:
Replaced with `awu.foreach`.
If two types have the same subtype, we assume that the same fields of the subtype are deployable in both types (which currently seems correct for Jira). If that's not the case, we will need to separate the subtype to multiple subtypes

netama reviewed (commented):
added some comments and questions - I think it would be better to call out the various changes happening here in the PR description, as there are some that I think are different from the previous design :)
specifically - 
1. IIUC it looks like we changed the default for deployable fields from true to false? why was that? 
2. I think there are some potential race conditions with the new `Promise.all()` calls - I'm not sure I understand if there is a real risk or not, but if there isn't I think we should enforce it
3. are we covering `additionalProperties` with this?

alonstern commented:
> added some comments and questions - I think it would be better to call out the various changes happening here in the PR description, as there are some that I think are different from the previous design :) specifically -
> 
> 1. IIUC it looks like we changed the default for deployable fields from true to false? why was that? won't it be much more noisy (even if only in the state)?
> 2. I think there are some potential race conditions with the new `Promise.all()` calls - I'm not sure I understand if there is a real risk or not, but if there isn't I think we should enforce it
> 3. are we covering `additionalProperties` with this?

1. The default was not changed, I only changed it to always write true, false and not leave the annotation empty, but that's not really important so I changed it
2. Fixed
3. Fixed

netama commented on packages/adapter-components/src/elements/swagger/deployment/annotations.ts:
```diff
@@ -13,49 +13,92 @@
+    }
+
+    if (isMapType(fieldType)) {
+      const innerType = await fieldType.getInnerType()
```
I think both here and below this might need to be `getDeepInnerType` (for maps-of-lists etc)

netama reviewed (approved):
LGTM with one last nit
